sql: improve syntax around computed columns

justinj opened the pull request:
This commit makes two distinct but related changes:

* Generation expression for computed columns must be wrapped in
  parentheses.
* Parsing now provides more helpful error messages for computed columns.

Release note (sql change): Change computed column syntax and improved
error messages.

Commit:
sql: improve syntax around computed columns

This commit makes two distinct but related changes:

* Generation expression for computed columns must be wrapped in
  parentheses.
* Parsing now provides more helpful error messages for computed columns.

Release note (sql change): Change computed column syntax and improved
error messages.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22429)
<!-- Reviewable:end -->

BramGruneir commented:
A few comments.  Requiring extra parens around parens seems unwieldy and excessive.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 189 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVQSpV6kEVqP55Meb:-L4gVQSpV6kEVqP55Mec:bhlylsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L189)):*
> ```
>   a INT AS (3) VIRTUAL
> )
> ```

Can you add a test with extra unneeded parens?

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVDAtr42G9jplDmRM:-L4gVDAtr42G9jplDmRN:b-8hewpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L287)):*
> ```
>   k INT PRIMARY KEY,
>   a JSON,
>   b TEXT AS ((a->>'q')) STORED,
> ```

Is the 2nd set of parens needed here?

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 319 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVLMJQLkNLAQ3Sfs_:-L4gVLMJQLkNLAQ3Sfsa:b-8hewpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L319)):*
> ```
> statement ok
> CREATE TABLE x (
>   k INT AS (((data->>'id')::INT)) STORED PRIMARY KEY,
> ```

Is the 2nd set of parens needed here?

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 362 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVOQ5ZoSSmmuVgERB:-L4gVOQ5ZoSSmmuVgERC:b-8hewpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L362)):*
> ```
>   b INT,
>   c INT,
>   d INT[] AS ((ARRAY[a, b, c])) STORED
> ```

Is the 2nd set of parens needed here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-:-L4gVf_LMk3gSUupgnAf:b-isyojd)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 189 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVQSpV6kEVqP55Meb:-L4gWjozMili_RFpZzSG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L189)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add a test with extra unneeded parens?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVDAtr42G9jplDmRM:-L4gWdYuhxNZJbNRJElf:b-lnsvmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L287)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is the 2nd set of parens needed here?
</blockquote></details>

Changed from `b_expr` to `a_expr` to fix this.

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 319 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVLMJQLkNLAQ3Sfs_:-L4gWgYZZcXUj2fiT_Ea:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L319)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is the 2nd set of parens needed here?
</blockquote></details>

fixed

---

*[pkg/sql/logictest/testdata/logic_test/computed, line 362 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gVOQ5ZoSSmmuVgERB:-L4gWi3O99_TY3q58otA:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/logictest/testdata/logic_test/computed#L362)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is the 2nd set of parens needed here?
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 6 of 6 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/sql.y, line 3044 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gbDxvq-ApG6S3EIU6:-L4gbDxwj-VR2g7l9ST2:b-2jwa7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/parser/sql.y#L3044)):*
> ```Yacc
> //   REFERENCES <tablename> [( <colnames...> )] [ON DELETE {NO ACTION | RESTRICT}] [ON UPDATE {NO ACTION | RESTRICT}]
> //   COLLATE <collationname>
> //   AS <expr> STORED
> ```

Update this

---

*[pkg/sql/parser/sql.y, line 3347 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gbM387YAvTZIx9tWb:-L4gbM387YAvTZIx9tWc:becxt5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/parser/sql.y#L3347)):*
> ```Yacc
>     return unimplemented(sqllex, "virtual computed columns")
>  }
> | AS '(' b_expr ')' error
> ```

remove this rule entirely in favor of the one below

---

*[pkg/sql/parser/sql.y, line 3354 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gbP1PGusrGL9Y2-2J:-L4gbP1Quf3hff0wTy76:bq3zmmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/parser/sql.y#L3354)):*
> ```Yacc
> | AS error
>  {
>     sqllex.Error("syntax error near computed column - generation expression must be wrapped in parentheses")
> ```

make this error a reminder of the syntax: `syntax error: use AS ( <expr> ) [ STORED | VIRTUAL ]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 5 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/parser/sql.y, line 3044 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gbDxvq-ApG6S3EIU6:-L4gl0-nLlvC9yTOcOze:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/parser/sql.y#L3044)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Update this
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 3347 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gbM387YAvTZIx9tWb:-L4gl0hdtE8R4VSD8h7m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/parser/sql.y#L3347)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

remove this rule entirely in favor of the one below
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 3354 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-L4gbP1PGusrGL9Y2-2J:-L4gl1OqnBkcPnV5ghQh:b-j29s3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/8122a8eed91222a4ca7e2e3958f14a7308800843/pkg/sql/parser/sql.y#L3354)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make this error a reminder of the syntax: `syntax error: use AS ( <expr> ) [ STORED | VIRTUAL ]`
</blockquote></details>

Done in the way we discussed offline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22429#-:-L4gmA_bVu_2PU1JqE-w:bnfp4nl)*
<!-- Sent from Reviewable.io -->

justinj commented:
TFTRs!
